### PR TITLE
Add io.botfather.botfather

### PR DESCRIPTION
**Pull request by DoctorJohn**
Hey! Botfather is an universal automation application I made. Lately it got quite popular among students and I decided to improve the way I distribute the application.

This is my first app submission to flathub. I followed the "App Submission" guide and the "app requirements" document.

The application is not opensource yet, so I put binaries prebuild using flatpak-builder in the repositiory.
Please let me know whether something has to be changed in order for this application to get "accepted".

Thanks

### CODE REVIEW

**Comment by TingPing** [comment]
> The application is not opensource yet, so I put binaries prebuild using flatpak-builder in the repositiory.

git is not a place to store binaries. You should have releases hosted by the project.

File: cef/cef-3.3325.1758-linux64.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+	"name": "CEF",
+	"buildsystem": "cmake",
```
Comment (TingPing) [inline cef/cef-3.3325.1758-linux64.json]:
```suggestion
	"buildsystem": "cmake-ninja",
```

File: io.botfather.Botfather.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+<!-- Copyright 2018 The Botfather Developers -->
+<component type="desktop-application">
+	<id>io.botfather.Botfather.desktop</id>
+	<metadata_license>CC0-1.0</metadata_license>
+	<project_license>LicenseRef-proprietary=https://botfather.io/docs/legal/tos/</project_license>
+	<url type="homepage">https://botfather.io</url>
+	<name>Botfather</name>
+	<summary>Universal automation framework</summary>
+	
+	<categories>
```
Comment (TingPing) [inline io.botfather.Botfather.appdata.xml]: You can remove these, they go in the desktop file.

File: io.botfather.Botfather.appdata.xml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+			<image>https://botfather.io/static/metadata/flatpak/android-window.png</image>
+		</screenshot>
+		<screenshot>
+			<caption>Window showing the integrated browser</caption>
+			<image>https://botfather.io/static/metadata/flatpak/browser-window.png</image>
+		</screenshot>
+	</screenshots>
+
+	<content_rating type="oars-1.0">
+		<content_attribute id="money-purchasing">mild</content_attribute>
```
Comment (TingPing) [inline io.botfather.Botfather.appdata.xml]: So does it ask you to pay on first start or what?

File: io.botfather.Botfather.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  - "--share=ipc" # share IPC namespace with the host (necessary for X11)
+  - "--device=dri" # OpenGL rendering (for CEF)
+  - "--socket=pulseaudio" # play sound with PulseAudio
+  - "--share=network" # access the network (CEF)
+  - "--device=all" # for /dev/uinput (mouse and keyboard input)
+  - "--talk-name=org.freedesktop.ScreenSaver"
+build-options:
+  env:
+    CEF_ROOT: "/app/CEF"
+    LD_LIBRARY_PATH: "/app/lib"
```
Comment (TingPing) [inline io.botfather.Botfather.yaml]: This *shouldn't* be needed. If so I think its a runtime bug.

File: opencv/opencv-3.4.3.json [hunk 5]
```diff
@@ -0,0 +1,48 @@
+{
+	"name": "opencv",
+	"buildsystem": "cmake",
```
Comment (TingPing) [inline opencv/opencv-3.4.3.json]:
```suggestion
	"buildsystem": "cmake-ninja",
```

File: io.botfather.Botfather.appdata.xml (hunk 3 above)
Comment (DoctorJohn) [reply io.botfather.Botfather.appdata.xml]: No. One can use the software without paying for it.
To unlock extra functionality one has to log in using an account created on our website, which is eventually linked to Patreon.

**Comment by DoctorJohn** [comment]
> git is not a place to store binaries. You should have releases hosted by the project.

> Are these downloads usable: https://botfather.io/downloads/
>
> We might want this to become an extra-data source type. Currently I don't believe we host any proprietary apps and if we should seemed unclear.

Yeah, those downloads work. But I want to replace all linux builds with flatpak.

File: io.botfather.Botfather.yaml (hunk 4 above)
Comment (DoctorJohn) [reply io.botfather.Botfather.yaml]: They are indeed not needed. They are leftovers from the manifest I use to build my (prebuilt) binaries.

File: io.botfather.Botfather.appdata.xml (hunk 2 above)
Comment (DoctorJohn) [reply io.botfather.Botfather.appdata.xml]: Including the categories?

Comment (TingPing) [reply io.botfather.Botfather.appdata.xml]: Yes the `.desktop` file has a categories entry, they don't need to be in both files.

**Comment by DoctorJohn** [comment]
I eventually had to make CEF extra-data aswell, as our binaries are required to be next to the CEF resources.
@TingPing could you take another look? Thanks for your help and guidance so far

File: io.botfather.Botfather.appdata.xml [hunk 6]
```diff
@@ -5,14 +5,7 @@
 	<metadata_license>CC0-1.0</metadata_license>
 	<project_license>LicenseRef-proprietary=https://botfather.io/docs/legal/tos/</project_license>
 	<url type="homepage">https://botfather.io</url>
-	<name>Botfather</name>
```
[the remark this answers is not in the document]
Comment (DoctorJohn) [reply io.botfather.Botfather.appdata.xml]: Githubs code highlighting confused me there. I added the name again.

File: io.botfather.Botfather.yaml [hunk 7]
```diff
@@ -0,0 +1,63 @@
+app-id: io.botfather.Botfather
+branch: stable
+runtime: org.kde.Platform
+runtime-version: '5.11'
+sdk: org.kde.Sdk
+command: botfather
+tags:
+  - proprietary
+finish-args:
+  - "--filesystem=home:rw" # let user use their scripts properly
```
Comment (barthalion) [inline io.botfather.Botfather.yaml]: Shouldn't portal allow secure reads and writes?

Comment (DoctorJohn) [reply io.botfather.Botfather.yaml]: My understanding of portals was that I could use Qt methods like QFileDialog::getOpenFileName without having to worry about permissions at all.

I might got it wrong, but doing so didn't work for me. The QFileDialog doesn't allow the user to select anything within his home folder.

Furthermore the program can be scripted by the user to automatically generate images (not only screenshots) and save them using QImage::save. As far as I know QImage::save is not covered by portals, as it doesn't prompt the user.

**Comment by nedrichards** [comment]
bot, build io.botfather.Botfather

File: io.botfather.Botfather.yaml (hunk 7 above)
Comment (DoctorJohn) [reply io.botfather.Botfather.yaml]: I took another look at it. For the time being the application requires rw permissions to the to work as intended. That might change in the future though.

Details: A script file tells the program how to automate something. Those script files often come with images and other data i.e. for image recognition purposes. While running a script such data is read by the program. By introducting a convenient way to store scripts in a central location we might get rit of the home rw permission eventually.

**Comment by DoctorJohn** [comment]
Are there any more questions/issues?

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/io.botfather.Botfather

**Pull request closed without merging**